### PR TITLE
Update how RMSE and MUE (and their CIs) are computed

### PR DESCRIPTION
**Pull request by zhang-ivy**
## Description
Fixes https://github.com/OpenFreeEnergy/cinnabar/issues/73 -- where in some cases, the RMSE and MUE does not lie within the 95% CI

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] Change the default bootstrapping mode to only account for finite sample size effects (tells us how well we fit the data)
  - [x] Support the old bootstrapping mode (which also accounts for simulating what would happen if we ran the experiment and/or calculation many times) by allowing the user to separately control whether experimental and predicted uncertainty should be included when bootstrapping

<strike>  - [x] For both modes, instead of reporting the MLE of RMSE and MUE, report the mean of the bootstrapped data</strike>
  - [x] For the default mode, continue to report the sample estimate of the RMSE and MUE. For the old mode, report the mean of the bootstrapped data

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by zhang-ivy** [comment]
@IAlibay : Done! Let me know if you think there is a better test to add

File: cinnabar/tests/test_stats.py [hunk 1]
```diff
@@ -145,3 +145,24 @@ def test_correlation_positive(fe_map):
         assert (
             0.5 < bss["mle"] < 0.9
         ), f"Correlation must be positive for this data. {stat} is {bss['mle']}"
+
```
Comment (IAlibay) [inline cinnabar/tests/test_stats.py:148]:
```suggestion


```
PEP8

File: cinnabar/tests/test_stats.py [hunk 2]
```diff
@@ -145,3 +145,24 @@ def test_correlation_positive(fe_map):
...
+def test_confidence_intervals(fe_map):
+    """
+    Test that the confidence intervals for RMSE and MUE contain the corresponding statistics.
+    Uses the example data in `cinnabar/data/example.csv`
```
Comment (IAlibay) [inline cinnabar/tests/test_stats.py:152]:
```suggestion
    Test that the confidence intervals for RMSE and MUE contain the
    corresponding statistics. Uses the example data in
    `cinnabar/data/example.csv`
```
PEP8

File: cinnabar/tests/test_stats.py [hunk 3]
```diff
@@ -145,3 +145,24 @@ def test_correlation_positive(fe_map):
...
+    nodes = fe_map.graph.nodes
+
+    x_data = np.asarray([n[1]["exp_DG"] for n in nodes(data=True)])
+    y_data = np.asarray([n[1]["calc_DG"] for n in nodes(data=True)])
+    xerr = np.asarray([n[1]["exp_dDG"] for n in nodes(data=True)])
+    yerr = np.asarray([n[1]["calc_dDG"] for n in nodes(data=True)])
+
+    # RMSE
+    bss = bootstrap_statistic(x_data, y_data, xerr, yerr, statistic="RMSE")
+    assert bss['low'] < bss['mean'] < bss['high'], "The RMSE must lie withint the bootstrapped 95% CI"
```
Comment (IAlibay) [inline cinnabar/tests/test_stats.py:164]: This is a good first test but it doesn't actually trigger either of the newly introduced bools for adding noise either in the x or y direction. Ideally we need a test for that, probably maybe a regression test (assuming enough bootstraps to ensure convergence)?

File: cinnabar/stats.py [hunk 4]
```diff
@@ -14,6 +14,8 @@ def bootstrap_statistic(
     statistic: str = "RMSE",
     nbootstrap: int = 1000,
     plot_type: str = "dG",
+    bootstrap_experimental_uncertainty: bool = False,
```
Comment (IAlibay) [inline cinnabar/stats.py:17]: I would prefer we use the labels `x_true` and `y_pred` for these in the same manner as the other arguments for the method. Whilst the normal use is experiment & predicted, there is no reason why this can't be predicted1 vs predicted2, or even experiment1 vs experiment2.

**Review by IAlibay — CHANGES_REQUESTED** [review]
Sorry I don't think I will have time today to dig into what would make an ideal test for this, I can try to come back to it tomorrow if needed.

If possible could you check for other PEP8 issues in the newly added code please? Specifically lines 164 and 168 seem to exceed 79 chars.

File: cinnabar/stats.py (hunk 4 above)
Comment (zhang-ivy) [reply cinnabar/stats.py]: Fixed!

File: cinnabar/plotting.py [hunk 5]
```diff
@@ -167,8 +172,14 @@ def _master_plot(
...
-        s = stats.bootstrap_statistic(x, y, xerr, yerr, statistic=statistic)
-        string = f"{statistic}:   {s['mle']:.2f} [95%: {s['low']:.2f}, {s['high']:.2f}] " + "\n"
+        s = stats.bootstrap_statistic(x,
+                                      y,
+                                      xerr,
+                                      yerr,
+                                      statistic=statistic,
+                                      bootstrap_experimental_uncertainty=bootstrap_experimental_uncertainty,
+                                      bootstrap_predicted_uncertainty=bootstrap_predicted_uncertainty)
+        string = f"{statistic}:   {s['mean']:.2f} [95%: {s['low']:.2f}, {s['high']:.2f}] " + "\n"
```
Comment (IAlibay) [inline cinnabar/plotting.py:182]: Just an additional thought here, by using 'mean' for all cases here we do slightly move away from what's in the original issue (I think?).

Now we're saying that in all cases the estimate from the bootstrap samples rather than the original sample estimate is the thing we report. Assuming a statistic that is a linear function[note1] of the original data and enough bootstraps, I assume the two will come close but they are unlikely to be the same?

My understanding is that folks preferred using the sample estimate where possible (i.e. the case of adding noise is one of those cases where it's not possible). I'm not sure I have a formed opinion yet on which is better to be the most reported value[note2], that being said I do believe it to be important that we at least warn users of the difference between the two estimates, ensuring that folks are aware of any bias in the estimates.

[note1]: Are we guaranteeing this?
[note2]: I think I take the stance that ideally both would be reported when communicating science, but I do know it gets messy.

Comment (zhang-ivy) [reply cinnabar/plotting.py:182]: >My understanding is that folks preferred using the sample estimate where possible (i.e. the case of adding noise is one of those cases where it's not possible).

I was unsure of this as well -- I thought @jchodera had mentioned in the call yesterday that in both the default and old modes, we want to use the mean instead of the MLE, but now that I'm thinking about it, he might have meant that when using either of the newly introduced bools are used, we want to use the mean. 

I will update the code so that in the default mode, the sample estimate (`s['MLE']`) is used. If one of the newly introduced bools is used, we used mean (`s['mean']`)

File: cinnabar/stats.py (hunk 4 above)
Comment (ijpulidos) [reply cinnabar/stats.py]: I agree with this suggestion, it is better if we have names that tell more. Maybe something that  has `true` and `predicted` on the names, `include_true_errors` and `include_predicted_errors` or something similar. By the name of the function we already know this is related to boostrapping, so it feels unnecessary to have it in the `kwarg` name.

File: cinnabar/plotting.py (hunk 5 above)
Comment (zhang-ivy) [reply cinnabar/plotting.py]: Just pushed the changes

Comment (IAlibay) [reply cinnabar/plotting.py]: If @jchodera has the time to review this it would be grand, it is one of those things where it seems like an innocuous change, but I do worry it will have wider consequences than I can see.

Maybe an opinion by @ppxasjsm would be good? That way we're a bit more sure we aren't messing up other's pipelines?

File: cinnabar/stats.py (hunk 4 above)
Comment (zhang-ivy) [reply cinnabar/stats.py]: Just changed `bootstrap` to `include` for the arguments to `bootstrap_statistics`. Did not change the argument names plotting.py and plotlying.py functions

File: cinnabar/tests/test_stats.py (hunk 3 above)
Comment (zhang-ivy) [reply cinnabar/tests/test_stats.py]: I just changed the test to use the newly introduced bools and check whether the RMSE lies within the 95% CIs. 

I thought about adding a regression test -- was thinking I could add my data from #73 to `cinnabar/data/` and then create a test that checks whether that specific dataset passes the test, but I'm not using a standard cinnabar csv input file for that dataset, because its not the standard "predicted DDG" vs "experiment DDG", its "forward DDG" vs "reverse DDG." It would take a bit of time to properly format it and I'm not sure its essential to include, so I decided not to.

Comment (zhang-ivy) [reply cinnabar/tests/test_stats.py]: Let me know if you think of better tests / feel free to add them yourself if you have time!

File: cinnabar/plotting.py [hunk 6]
```diff
@@ -167,8 +172,15 @@ def _master_plot(
...
-        s = stats.bootstrap_statistic(x, y, xerr, yerr, statistic=statistic)
-        string = f"{statistic}:   {s['mle']:.2f} [95%: {s['low']:.2f}, {s['high']:.2f}] " + "\n"
+        s = stats.bootstrap_statistic(x,
+                                      y,
+                                      xerr,
+                                      yerr,
+                                      statistic=statistic,
+                                      include_true_uncertainty=bootstrap_x_uncertainty,
+                                      include_pred_uncertainty=bootstrap_y_uncertainty)
+        statistic_type = 'mean' if bootstrap_x_uncertainty or bootstrap_y_uncertainty else 'mle'
```
Comment (jchodera) [inline cinnabar/plotting.py:182]: Instead of auto-selecting the mean or mle depending on whether bootstrap_{x|y}_uncertainty is specified, can we just have a separate kwarg for `statistic_type` that defaults to `mle` and allows the user to set it to `mean` if they prefer? That would make the behavior much more obvious.

**Review by jchodera — CHANGES_REQUESTED** [review]
There's one minor change that would help reduce the surprise developers will have in using this, but other than this, it looks great!

**Comment by zhang-ivy** [comment]
@IAlibay @ijpulidos : could you review this again / merge if it looks good?

File: cinnabar/plotting.py [hunk 7]
```diff
@@ -342,6 +373,9 @@ def plot_DGs(
     plotly: bool = False,
     centralizing: bool = True,
     shift: float = 0.0,
+    bootstrap_x_uncertainty: bool = False,
+    bootstrap_y_uncertainty: bool=False,
```
Comment (ijpulidos) [inline cinnabar/plotting.py:377]: Just minor style fix.
```suggestion
    bootstrap_y_uncertainty: bool = False,
```

**Review by ijpulidos — APPROVED** [review]
Looks good to me! Just a minor style change. Not blocker.

File: cinnabar/plotting.py (hunk 7 above)
Comment (zhang-ivy) [reply cinnabar/plotting.py]: Good catch, thanks!

File: cinnabar/plotting.py [hunk 8]
```diff
@@ -87,7 +90,12 @@ def _master_plot(
...
     font_sizes : dict, default {"title": 12, "labels": 9, "other": 12}
         font sizes to use for the title ("title"), the data labels ("labels"), and the rest of the plot ("other")
-
+    bootstrap_x_uncertainty : bool, default False
+        whether to account for uncertainty in x when bootstrapping
+    bootstrap_y_uncertainty : bool, default False
+        whether to account for uncertainty in y when bootstrapping
+    statistic_type : str, default 'mle'
+        the type of statistic to use, either 'mle' or 'mean' (of the bootstrapped data)
     Returns
```
Comment (IAlibay) [inline cinnabar/plotting.py:99]:
```suggestion

    Returns
```

File: cinnabar/plotlying.py [hunk 9]
```diff
@@ -266,11 +269,18 @@ def _master_plot(
 
     # stats and title
     string = []
+    if statistic_type not in ['mle', 'mean']:
+        raise Exception(f"Unknown statistic type {statistic_type}")
```
Comment (IAlibay) [inline cinnabar/plotlying.py:273]:
```suggestion
        raise ValueError(f"Unknown statistic type {statistic_type}")
```
Please don't raise a plain Exception, use a `ValueError` here? Also please add test for exception.

File: cinnabar/plotting.py [hunk 10]
```diff
@@ -166,9 +174,17 @@ def _master_plot(
 
     # stats and title
     statistics_string = ""
+    if statistic_type not in ['mle', 'mean']:
+        raise Exception(f"Unknown statistic type {statistic_type}")
```
Comment (IAlibay) [inline cinnabar/plotting.py:178]:
```suggestion
        raise ValueError(f"Unknown statistic type {statistic_type}")
```
As above + test

File: cinnabar/plotting.py [hunk 11]
```diff
@@ -358,6 +392,12 @@ def plot_DGs(
         Title for the plot
     filename : str, default = None
         filename for plot
+    bootstrap_x_uncertainty : bool, default False
+        whether to account for uncertainty in x when bootstrapping
+    bootstrap_y_uncertainty : bool, default False
+        whether to account for uncertainty in y when bootstrapping
+    statistic_type : str, default 'mle'
+        the type of statistic to use, either 'mle' or 'mean' (of the bootstrapped data)
```
Comment (IAlibay) [inline cinnabar/plotting.py:400]: As above

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: cinnabar/plotting.py [hunk 12]
```diff
@@ -87,7 +90,12 @@ def _master_plot(
         contains the minimium and maximum values to use for the x and y axes. if specified, axis_padding is ignored
     font_sizes : dict, default {"title": 12, "labels": 9, "other": 12}
         font sizes to use for the title ("title"), the data labels ("labels"), and the rest of the plot ("other")
-
+    bootstrap_x_uncertainty : bool, default False
+        whether to account for uncertainty in x when bootstrapping
+    bootstrap_y_uncertainty : bool, default False
+        whether to account for uncertainty in y when bootstrapping
+    statistic_type : str, default 'mle'
+        the type of statistic to use, either 'mle' or 'mean' (of the bootstrapped data)
```
[the remark this answers is not in the document]
Comment (zhang-ivy) [reply cinnabar/plotting.py]: Done

File: cinnabar/plotting.py [hunk 13]
```diff
@@ -235,6 +254,12 @@ def plot_DDGs(
             the negative of the transformation
         currently unsupported for plotly-generated plots
         TODO: implement data labeling for the case where plotly=True
+    bootstrap_x_uncertainty : bool, default False
+        whether to account for uncertainty in x when bootstrapping
+    bootstrap_y_uncertainty : bool, default False
+        whether to account for uncertainty in y when bootstrapping
+    statistic_type : str, default 'mle'
+        the type of statistic to use, either 'mle' or 'mean' (of the bootstrapped data)
```
[the remark this answers is not in the document]
Comment (zhang-ivy) [reply cinnabar/plotting.py]: Done

File: cinnabar/plotting.py (hunk 11 above)
Comment (zhang-ivy) [reply cinnabar/plotting.py]: Done

File: cinnabar/plotting.py [hunk 14]
```diff
@@ -437,6 +486,12 @@ def plot_all_DDGs(
         whether to use plotly for the plotting
     shift : float, default = 0.
         shift both the x and y axis by a constant
+    bootstrap_x_uncertainty : bool, default False
+        whether to account for uncertainty in x when bootstrapping
+    bootstrap_y_uncertainty : bool, default False
+        whether to account for uncertainty in y when bootstrapping
+    statistic_type : str, default 'mle'
+        the type of statistic to use, either 'mle' or 'mean' (of the bootstrapped data)
```
[the remark this answers is not in the document]
Comment (zhang-ivy) [reply cinnabar/plotting.py]: Done

File: cinnabar/tests/test_stats.py [hunk 15]
```diff
@@ -145,3 +145,47 @@ def test_correlation_positive(fe_map):
...
+    nodes = fe_map.graph.nodes
+
+    x_data = np.asarray([n[1]["exp_DG"] for n in nodes(data=True)])
+    y_data = np.asarray([n[1]["calc_DG"] for n in nodes(data=True)])
+    xerr = np.asarray([n[1]["exp_dDG"] for n in nodes(data=True)])
+    yerr = np.asarray([n[1]["calc_dDG"] for n in nodes(data=True)])
+
+    # RMSE (default mode)
+    bss = bootstrap_statistic(x_data, y_data, xerr, yerr, statistic="RMSE")
+    assert bss['low'] < bss['mle'] < bss['high'], \
```
[the remark this answers is not in the document]
Comment (zhang-ivy) [reply cinnabar/tests/test_stats.py]: Done

File: cinnabar/tests/test_stats.py [hunk 16]
```diff
@@ -145,3 +145,47 @@ def test_correlation_positive(fe_map):
...
+        "The RMSE must lie within the bootstrapped 95% CI"
+
+    # MUE (default mode)
+    bss = bootstrap_statistic(x_data, y_data, xerr, yerr, statistic="MUE")
```
[the remark this answers is not in the document]
Comment (zhang-ivy) [reply cinnabar/tests/test_stats.py:170]: @IAlibay : I'm not familiar with `pytest.parametrize` -- would you have time to re-format this with `pytest.parametrize`? If not, maybe we could just merge this PR as is and I could open an issue about improving the tests in the long term? While I'm sure your suggestion would improve the code, I'm not sure its essential, as my goal here was just to introduce the minimal changes necessary so that I (and others in the Chodera lab) can continue making progress on the scientific side of things.

File: cinnabar/tests/test_stats.py [hunk 17]
```diff
@@ -145,3 +145,47 @@ def test_correlation_positive(fe_map):
...
+        "The RMSE must lie within the bootstrapped 95% CI"
+
+    # RMSE (including yerr in bootstrapping)
+    bss = bootstrap_statistic(x_data, y_data, xerr, yerr, statistic="RMSE",
+                              include_pred_uncertainty=True)
+    assert bss['low'] < bss['mean'] < bss['high'], \
+        "The RMSE must lie within the bootstrapped 95% CI"
+
+    # RMSE (including both xerr and yerr in bootstrapping)
+    bss = bootstrap_statistic(x_data, y_data, xerr, yerr, statistic="RMSE",
```
[the remark this answers is not in the document]
Comment (zhang-ivy) [reply cinnabar/tests/test_stats.py:184]: > We're not testing passing statistic_type through
@IAlibay : That's because `statistic_type` is not an argument that's passed to `bootstrap_statistic`, it is only passed through `plotting.py` functions. We currently don't test any `plotting.py` functions, so would it be ok if I just opened an issue about writing tests for `plotting.py` as a long term action item?

[the remark this answers is not in the document]
Comment (zhang-ivy) [reply cinnabar/tests/test_stats.py:184]: > Is our dataset such that these could ever fail? Can we add the dataset from the problematic example (i.e. https://github.com/OpenFreeEnergy/cinnabar/issues/73 ) to make sure we test against that edge case properly?

I added a test addressing the case in my latest commit. I hard coded the value directly into the test (which @dotsdl mentioned would be ok), since I was having trouble formatting the cinnabar input CSV to properly handle my dataset (previously described this problem here: https://github.com/OpenFreeEnergy/cinnabar/pull/74#discussion_r1073863391)

[the remark this answers is not in the document]
Comment (IAlibay) [reply cinnabar/tests/test_stats.py:184]: >  We currently don't test any plotting.py functions, so would it be ok if I just opened an issue about writing tests for plotting.py as a long term action item?

Sorry that's on me, I should have realized that. Happy to go with just raising an issue about testing the plotting methods.

**Review by IAlibay — APPROVED** [review]

File: cinnabar/tests/test_stats.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (IAlibay) [reply cinnabar/tests/test_stats.py]: Done

File: cinnabar/tests/test_stats.py [hunk 18]
```diff
@@ -145,3 +146,71 @@ def test_correlation_positive(fe_map):
...
+    x_data = [-0.101, 0.351, 0.117, 0.623, 5.172, 5.209, -1.727, -1.387, -1.534, 1.082]
+    y_data = [-0.174, 0.42, 0.262, 0.626, 5.064, 4.783, -1.58, -1.712, -1.699, 0.822]
+    xerr = [0.443, 0.652, 0.57, 0.245, 1.112, 1.049, 1.23, 1.435, 1.521, 0.505]
+    yerr = [0.442, 0.714, 0.619, 0.224, 1.401, 1.107, 1.178, 1.252, 1.265, 0.472]
+
+    # RMSE (default mode)
+    bss = bootstrap_statistic(x_data, y_data, xerr, yerr, statistic="RMSE",
+                              include_true_uncertainty=True,
+                              include_pred_uncertainty=True)
+    assert (bss['low'] > bss['mle']) or (bss['mle'] > bss['high']), error_message
```
Comment (zhang-ivy) [inline cinnabar/tests/test_stats.py:216]: Was reviewing @IAlibay 's changes and just realized that `error_message` isn't defined here. I'll open a PR to fix this.

**Comment by zhang-ivy** [comment]
Thanks for your help and feedback @IAlibay ! Will try to incorporate pytest fixtures / parametrize in the future

File: cinnabar/tests/test_stats.py (hunk 18 above)
Comment (zhang-ivy) [reply cinnabar/tests/test_stats.py:216]: Here's the PR: https://github.com/OpenFreeEnergy/cinnabar/pull/76
Note that I also found another issue with the edge case test and fixed it in the above PR
